File: examples/false_sharing.c
// False Sharing - Multi-threaded Cache Contention
// Two threads updating adjacent data cause cache line ping-pong
#include <stdio.h>
#include <pthread.h>

#ifndef ITERATIONS
#define ITERATIONS 1000000
#endif

// BAD: Both counters on same cache line (64 bytes)
struct {
    volatile int counter1;
    volatile int counter2;
} shared_bad;

// GOOD: Counters on separate cache lines
struct {
    volatile int counter1;
    char padding[60];  // Pad to 64 bytes
    volatile int counter2;
} shared_good;

void* increment_counter1(void* arg) {
    for (int i = 0; i < ITERATIONS; i++) {
#ifdef RUN_PADDED
        shared_good.counter1++;
#else
        shared_bad.counter1++;  // Causes invalidation of counter2's cache line
#endif
    }
    return NULL;
}

void* increment_counter2(void* arg) {
    for (int i = 0; i < ITERATIONS; i++) {
#ifdef RUN_PADDED
        shared_good.counter2++;
#else
        shared_bad.counter2++;  // Causes invalidation of counter1's cache line
#endif
    }
    return NULL;
}

int main() {
    pthread_t t1, t2;

    // Create threads that update adjacent data
    pthread_create(&t1, NULL, increment_counter1, NULL);
    pthread_create(&t2, NULL, increment_counter2, NULL);

    pthread_join(t1, NULL);
    pthread_join(t2, NULL);

#ifdef RUN_PADDED
    printf("Padded Counter1: %d, Counter2: %d\n",
           shared_good.counter1, shared_good.counter2);
#else
    printf("Packed Counter1: %d, Counter2: %d\n",
           shared_bad.counter1, shared_bad.counter2);
#endif

    return 0;
}

File: examples/false_sharing.rs
// False sharing demonstration (Rust version)
// Expected: Cache invalidations when threads modify adjacent data
// Demonstrates: std::thread, AtomicUsize, cache line padding

use std::sync::atomic::{AtomicUsize, Ordering};
use std::thread;
use std::time::Instant;

const CACHE_LINE_SIZE: usize = 64;
const NUM_THREADS: usize = 4;
const ITERATIONS: usize = 100000;

// Bad: counters packed together, will cause false sharing
struct PackedCounters {
    counters: [AtomicUsize; NUM_THREADS],
}

impl PackedCounters {
    fn new() -> Self {
        Self {
            counters: std::array::from_fn(|_| AtomicUsize::new(0)),
        }
    }
}

// Good: counters padded to separate cache lines
#[repr(C, align(64))]
struct PaddedCounter {
    value: AtomicUsize,
    _padding: [u8; CACHE_LINE_SIZE - std::mem::size_of::<AtomicUsize>()],
}

impl PaddedCounter {
    fn new() -> Self {
        Self {
            value: AtomicUsize::new(0),
            _padding: [0; CACHE_LINE_SIZE - std::mem::size_of::<AtomicUsize>()],
        }
    }
}

struct PaddedCounters {
    counters: [PaddedCounter; NUM_THREADS],
}

impl PaddedCounters {
    fn new() -> Self {
        Self {
            counters: std::array::from_fn(|_| PaddedCounter::new()),
        }
    }
}

fn run_packed_test(counters: &PackedCounters) {
    let start = Instant::now();

    thread::scope(|s| {
        for i in 0..NUM_THREADS {
            s.spawn(move || {
                for _ in 0..ITERATIONS {
                    counters.counters[i].fetch_add(1, Ordering::Relaxed);
                }
            });
        }
    });

    let duration = start.elapsed();
    println!("Packed (false sharing): {:?}", duration);
}

fn run_padded_test(counters: &PaddedCounters) {
    let start = Instant::now();

    thread::scope(|s| {
        for i in 0..NUM_THREADS {
            s.spawn(move || {
                for _ in 0..ITERATIONS {
                    counters.counters[i].value.fetch_add(1, Ordering::Relaxed);
                }
            });
        }
    });

    let duration = start.elapsed();
    println!("Padded (no false sharing): {:?}", duration);
}

fn main() {
    println!("False Sharing Demonstration (Rust)");
    println!("Cache line size: {} bytes", CACHE_LINE_SIZE);
    println!("Threads: {}", NUM_THREADS);
    println!("Iterations per thread: {}", ITERATIONS);
    println!();

    let packed = PackedCounters::new();
    let padded = PaddedCounters::new();

    run_packed_test(&packed);
    run_padded_test(&padded);
}
